### PR TITLE
libcurl/all: Add CoreServices framework on macOS 14 Sonoma

### PR DESCRIPTION
**Pull request by nikolay-pv**
Specify library name and version:  **libcurl/all** on macOS (14 and up)

macOS 14 Sonoma has new framework CoreServices, to which libcurl stated linking in https://github.com/curl/curl/commit/6ab7e1990bd548059f08c471d20537bca13c67b6

This PR adds the `CoreServices` to the list of system frameworks only when building on macOS 14.

Absence of that framework caused a runtime crash earlier today to one of the apps as described in [here](https://github.com/curl/curl/issues/11893), however, I couldn't reproduce it with a provided test_package (or FWIW a tweaked one).

```
# conan --version
Conan version 2.0.14
```

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by valgur — APPROVED** [review]

**Comment by jcar87** [comment]
Hi @nikolay-pv - thank you for the PR and for linking the relevant information, very useful.

I can see that this seems to only be an issue when you explicitly disable `c-ares` support, could you confirm if this is the case?
On the other hand, I think that much like the libcurl fix, we _don't_ need to conditionally guard this with `os.version` - since the `CoreServices` framework is available in earlier versions as well, and `os.version` is not typically defined (by default by Conan)

**Comment by jcar87** [comment]
@nikolay-pv - I'm unable to reproduce this issue on macOS sonoma with any version of `libcurl` or AppleClang. The runtime linker seems to load `/System/Library/PrivateFrameworks/CoreServicesInternal.framework/Versions/A/CoreServicesInternal` regardless of whether it was passed in the command line as a linker flag or not. I'd be happy to move forward with the proposed change (for all versions of macOS), given the wealth of evidence - but there's also a chance that the bug was fixed in newer versions of Sonoma, or that it only manifests for specific codepaths?

**Comment by nikolay-pv** [comment]
Hi @jcar87, thanks a lot for taking the time to look into this change!

Somehow I thought CoreServices wasn't available in earlier macOS, but, indeed, it seems like it has been there for quite a long time.

I've re-checked and I still get the crash in the app we have, after looking around, I think I got down to a reproducible small sample. FWIF, I have macOS `14.2 (23C64)`, and `clang --version` gives:
> Apple clang version 15.0.0 (clang-1500.3.9.4)
> Target: arm64-apple-darwin23.2.0
> Thread model: posix

I have then a project folder with those three files (I pinned the dependencies to versions I could see in CMakeCache.txt):
[CMakeLists.txt](https://github.com/conan-io/conan-center-index/files/15250185/CMakeLists.txt)
[conanfile.txt](https://github.com/conan-io/conan-center-index/files/15250186/conanfile.txt)
[main_cpp.txt](https://github.com/conan-io/conan-center-index/files/15250192/main_cpp.txt)

To get the crash I run the following in a folder with them:
```sh
mkdir conan_output
cd conan_output
conan install .. --build missing
cd ..
cmake -S . -B build -G 'Ninja Multi-Config'
ninja -C build -f build-Debug.ninja all
./build/Debug/main
```
The output is then:
```
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[__NSCFString stringByStandardizingPath]: unrecognized selector sent to instance 0x1430045d0'
*** First throw call stack:
(
        0   CoreFoundation                      0x000000018268c570 __exceptionPreprocess + 176
        1   libobjc.A.dylib                     0x000000018217deb4 objc_exception_throw + 60
        2   CoreFoundation                      0x000000018273e12c -[NSObject(NSObject) __retain_OA] + 0
        3   CoreFoundation                      0x00000001825f67d4 ___forwarding___ + 1572
        4   CoreFoundation                      0x00000001825f60f0 _CF_forwarding_prep_0 + 96
        5   Foundation                          0x00000001836cc860 -[NSProcessInfo arguments] + 188
        6   CoreFoundation                      0x0000000182707d48 __getDefaultArguments_block_invoke + 96
        7   libdispatch.dylib                   0x0000000182388910 _dispatch_client_callout + 20
        8   libdispatch.dylib                   0x000000018238a14c _dispatch_once_callout + 32
        9   CoreFoundation                      0x00000001827076e4 _addBackstopValuesForIdentifierAndSource + 640
        10  CoreFoundation                      0x00000001825c1ffc __81-[_CFXPreferences(SourceAdditions) withNamedVolatileSourceForIdentifier:perform:]_block_invoke + 144
        11  CoreFoundation                      0x000000018270738c -[_CFXPreferences withNamedVolatileSourceForIdentifier:perform:] + 272
        12  CoreFoundation                      0x00000001825c836c -[CFPrefsSearchListSource addNamedVolatileSourceForIdentifier:] + 136
        13  CoreFoundation                      0x000000018274660c __108-[_CFXPreferences(SearchListAdditions) withSearchListForIdentifier:container:cloudConfigurationURL:perform:]_block_invoke.155 + 296
        14  CoreFoundation                      0x00000001827462b4 -[_CFXPreferences withSearchLists:] + 84
        15  CoreFoundation                      0x00000001825c38c0 __108-[_CFXPreferences(SearchListAdditions) withSearchListForIdentifier:container:cloudConfigurationURL:perform:]_block_invoke + 300
        16  CoreFoundation                      0x0000000182746460 -[_CFXPreferences withSearchListForIdentifier:container:cloudConfigurationURL:perform:] + 384
        17  CoreFoundation                      0x00000001825c31e8 -[_CFXPreferences copyAppValueForKey:identifier:container:configurationURL:] + 156
        18  CoreFoundation                      0x00000001825c3110 _CFPreferencesCopyAppValueWithContainerAndConfiguration + 112
        19  SystemConfiguration                 0x0000000183355478 SCDynamicStoreCopyProxiesWithOptions + 180
        20  main                                0x00000001008c7078 Curl_macos_init + 16
        21  main                                0x00000001008ad578 curl_global_init + 232
        22  main                                0x0000000100897d24 _ZNK3$_0clEv + 24
        23  main                                0x0000000100897d64 __cxx_global_var_init + 20
        24  main                                0x0000000100897d84 _GLOBAL__sub_I_main.cpp + 12
        25  dyld                                0x00000001821d5a24 ___ZZNK5dyld46Loader25findAndRunAllInitializersERNS_12RuntimeStateEENK3$_0clEv_block_invoke + 168
        26  dyld                                0x000000018221b328 ___ZNK5dyld313MachOAnalyzer18forEachInitializerER11DiagnosticsRKNS0_15VMAddrConverterEU13block_pointerFvjEPKv_block_invoke.209 + 340
        27  dyld                                0x000000018220e668 ___ZNK5dyld39MachOFile14forEachSectionEU13block_pointerFvRKNS0_11SectionInfoEbRbE_block_invoke + 496
        28  dyld                                0x00000001821b52fc _ZNK5dyld39MachOFile18forEachLoadCommandER11DiagnosticsU13block_pointerFvPK12load_commandRbE + 300
        29  dyld                                0x000000018220d6a0 _ZNK5dyld39MachOFile14forEachSectionEU13block_pointerFvRKNS0_11SectionInfoEbRbE + 192
        30  dyld                                0x000000018221ae3c _ZNK5dyld313MachOAnalyzer18forEachInitializerER11DiagnosticsRKNS0_15VMAddrConverterEU13block_pointerFvjEPKv + 516
        31  dyld                                0x00000001821d1b38 _ZNK5dyld46Loader25findAndRunAllInitializersERNS_12RuntimeStateE + 524
        32  dyld                                0x00000001821d7f70 _ZNK5dyld416JustInTimeLoader15runInitializersERNS_12RuntimeStateE + 36
        33  dyld                                0x00000001821d1f24 _ZNK5dyld46Loader23runInitializersBottomUpERNS_12RuntimeStateERN5dyld35ArrayIPKS0_EE + 220
        34  dyld                                0x00000001821d5ab0 _ZZNK5dyld46Loader38runInitializersBottomUpPlusUpwardLinksERNS_12RuntimeStateEENK3$_1clEv + 112
        35  dyld                                0x00000001821d20f0 _ZNK5dyld46Loader38runInitializersBottomUpPlusUpwardLinksERNS_12RuntimeStateE + 380
        36  dyld                                0x00000001821f74dc _ZN5dyld44APIs25runAllInitializersForMainEv + 464
        37  dyld                                0x00000001821b9fa0 _ZN5dyld4L7prepareERNS_4APIsEPKN5dyld313MachOAnalyzerE + 3192
        38  dyld                                0x00000001821b8edc start + 1844
)
libc++abi: terminating due to uncaught exception of type NSException
[1]    83809 abort      ./build/Debug/main
```
 
 If I then uncomment the linking to `CORE_SERVICES_FRAMEWORK` in CMakeLists.txt, the crash doesn't happen (the below lines):
 ```cmake
find_library(CORE_SERVICES_FRAMEWORK CoreServices REQUIRED)
target_link_libraries(main PRIVATE ${CORE_SERVICES_FRAMEWORK})
```

**Comment by jcar87** [comment]
Excellent! thank you so much, will get on this right away

**Comment by jcar87** [comment]
Thanks @nikolay-pv - I've been able to reproduce the issue with the code you have provided. Interesting that it happens when it is called like that, I had also tried the following, but was still succeeding:
* the test_package as it is, with `curl_easy_init` - which implicitly calls `curl_global_init`
* calling it like this, like in the documentation:
```
int main(void)
{
  curl_global_init(CURL_GLOBAL_DEFAULT);
 
  /* use libcurl, then before exiting... */
 
  curl_global_cleanup();
}
```

On the other hand I can see that the hooks were. already alerting us of this:
```
post_package(): WARN: [MISSING SYSTEM LIBS (KB-H043)] Library './lib/libcurl.4.dylib' links to system library 'CoreServices' but it is not in cpp_info.frameworks.
```

I've amended your PR to pass this regardless of macOS version, to match what curl has done upstream - you will need to fetch and reset if you want to continue working on this branch. Thanks a lot for providing additional information!

**Comment by nikolay-pv** [comment]
@jcar87 indeed, it is only when it is a call in the static initialisation context (we have a small RAII class in our code but initialise it in static context). 

Do I need to take any actions to get this PR merged in main?

File: recipes/libcurl/all/conanfile.py [hunk 1]
```diff
@@ -680,6 +680,8 @@ def package_info(self):
                 self.cpp_info.components["curl"].system_libs.append("crypt32")
         elif is_apple_os(self):
             self.cpp_info.components["curl"].frameworks.append("CoreFoundation")
+            if Version(self.settings.os.version) >= 14:
+                self.cpp_info.components["curl"].frameworks.append("CoreServices")
```
Comment (jcar87) [inline recipes/libcurl/all/conanfile.py:684]:
```suggestion
            self.cpp_info.components["curl"].frameworks.append("CoreServices")
```

**Review by jcar87 — APPROVED** [review]

**Review by franramirez688 — APPROVED** [review]

**Pull request merged**